girder-dandi-archive/girder_dandi_archive/locking.py

```diff
@@ -0,0 +1,143 @@
+    """
+    user_id = str(user["_id"]) if user is not None else None
+    return _get_lock(identifier) == user_id
+
+
+def require_lock(identifier, user):
+    """
+    Require that the user has the lock on the given dandiset.
+
+    If user is None, then it will verify that there is no lock.
+    This method should be used to ensure that a lock is present.
+    Note the different between this method and require_access.
+    """
+    if not has_lock(identifier, user):
+        locked_by = get_lock_owner(identifier)
+        if locked_by is not None:
+            owner_str = f"{locked_by['firstName']} {locked_by['lastName']}"
+            raise AccessException(f"Dandiset {identifier} is currently locked by {owner_str}")
+        else:
+            raise AccessException(f"Dandiset {identifier} is currently unlocked")
+
+
+def require_access(identifier, user):
+    """
+    Require that the user either have the lock, or no lock is present.
+
+    This method should be used to determine if a user can modify a dandiset.
+    """
+    if (not has_lock(identifier, None)) and (not has_lock(identifier, user)):
+        locked_by = get_lock_owner(identifier)
+        owner_str = f"{locked_by['firstName']} {locked_by['lastName']}"
+        raise AccessException(f"Dandiset {identifier} is currently locked by {owner_str}")
+
+
+def lock(identifier, user):
+    """
+    Lock the given dandiset to the given user.
+
+    The dandiset must be unlocked.
+    """
+    require_lock(identifier, None)
+    _set_lock(identifier, user)
+
+
+def unlock(identifier, user):
+    """
+    Unlock the given dandiset.
+
+    The given user must currently have the lock on the dandiset.
+    """
+    require_lock(identifier, user)
+    _remove_lock(identifier)
+
+
+def get_lock_owner(identifier):
+    """Return the user who currently has the lock on the resource, or None if there is no lock."""
+    lock_id = _get_lock(identifier)
+    if lock_id is None:
+        return None
+    return User().findOne(ObjectId(lock_id))
+
+
+def folder_save_listener(event):
+    """
+    Listen to folder save events to enforce locking.
+
+    If the folder has been locked by someone other than the current user,
+    an exception is thrown.
+    """
+    try:
+        identifier = event.info["meta"]["dandiset"]["identifier"]
+        require_access(identifier, getCurrentUser())
+    except KeyError:
+        # This event is on a non-dandiset folder, no locking is required
+        pass
+
+
+def upload_assetstore_listener(event):
+    """
+    Listen to upload assetstore events to enforce locking.
+
+    These events occur before files are uploaded.
+    If the root dandiset folder has been locked by someone other than the current user,
+    an exception is thrown.
+    """
+    thing = event.info["resource"]
+    while thing["parentCollection"] != "collection":
+        thing = Folder().findOne(thing["parentId"])
```

Comment:
I think some comments here would be useful, to describe what's going on to someone looking at it for the first time.